syntax = "proto3";

package envoy.service.ext_proc.v3alpha;

import "envoy/config/core/v3/base.proto";
import "envoy/extensions/filters/http/ext_proc/v3alpha/processing_mode.proto";
import "envoy/type/v3/http_status.proto";

import "google/protobuf/duration.proto";
import "google/protobuf/struct.proto";

import "udpa/annotations/status.proto";
import "validate/validate.proto";

option java_package = "io.envoyproxy.envoy.service.ext_proc.v3alpha";
option java_outer_classname = "ExternalProcessorProto";
option java_multiple_files = true;
option java_generic_services = true;
option (udpa.annotations.file_status).work_in_progress = true;
option (udpa.annotations.file_status).package_version_status = ACTIVE;

// [#protodoc-title: External Processing Service]

// A service that can access and modify HTTP requests and responses
// as part of a filter chain.
// The overall external processing protocol works like this:
//
// 1. Envoy sends to the service information about the HTTP request.
// 2. The service sends back a ProcessingResponse message that directs Envoy
//    to either stop processing, continue without it, or send it the
//    next chunk of the message body.
// 3. If so requested, Envoy sends the server chunks of the message body,
//    or the entire body at once. In either case, the server sends back
//    a ProcessingResponse after each message it receives.
// 4. If so requested, Envoy sends the server the HTTP trailers,
//    and the server sends back a ProcessingResponse.
// 5. At this point, request processing is done, and we pick up again
//    at step 1 when Envoy receives a response from the upstream server.
// 6. At any point above, if the server closes the gRPC stream cleanly,
//    then Envoy proceeds without consulting the server.
// 7. At any point above, if the server closes the gRPC stream with an error,
//    then Envoy returns a 500 error to the client, unless the filter
//    was configured to ignore errors.
//
// In other words, the process is a request/response conversation, but
// using a gRPC stream to make it easier for the server to
// maintain state.

service ExternalProcessor {
  // This begins the bidirectional stream that Envoy will use to
  // give the server control over what the filter does. The actual
  // protocol is described by the ProcessingRequest and ProcessingResponse
  // messages below.
  rpc Process(stream ProcessingRequest) returns (stream ProcessingResponse) {
  }
}

// This represents the different types of messages that Envoy can send
// to an external processing server.
// [#next-free-field: 8]
message ProcessingRequest {
  // Specify whether the filter that sent this request is running in synchronous
  // or asynchronous mode. If false, then the server must either respond
  // with exactly one ProcessingResponse message or close the stream.
  // If true, however, then the server must not respond with
  // an additional message, although it may still close the stream.
  // The choice of synchronous or asynchronous mode can be chosen in the
  // filter configuration.
  bool async_mode = 1;

  // Each request message will include one of the following sub-messages. Which
  // ones are set for a particular HTTP request/response depend on the
  // processing mode.
  oneof request {
    option (validate.required) = true;

    // Information about the HTTP request headers, as well as peer info and additional
    // properties. If "response_required" is set, the server must send back a
    // HeaderResponse message, an ImmediateResponse message, or close the stream.
    HttpHeaders request_headers = 2;

    // Information about the HTTP response headers, as well as peer info and additional
    // properties. If "response_required" is set, the server must send back a
    // HeaderResponse message or close the stream.
    HttpHeaders response_headers = 3;

    // A chunk of the HTTP request body. If "response_required" is set, the server must send back
    // a BodyResponse message, an ImmediateResponse message, or close the stream.
    HttpBody request_body = 4;

    // A chunk of the HTTP request body. If "response_required" is set, the server must send back
    // a BodyResponse message or close the stream.
    HttpBody response_body = 5;

    // The HTTP trailers for the request path. If "response_required" is set, the server
    // must send back a TrailerResponse message or close the stream.
    HttpTrailers request_trailers = 6;

    // The HTTP trailers for the response path. If "response_required" is set, the server
    // must send back a TrailerResponse message or close the stream.
    HttpTrailers response_trailers = 7;
  }
}

// For every ProcessingRequest received by the server with the "async_mode" field
// set to false, the server must send back exactly one ProcessingResponse message.
// [#next-free-field: 10]
message ProcessingResponse {
  oneof response {
    option (validate.required) = true;

    // The server must send back this message in response to a message with the
    // "request_headers" field set.
    HeadersResponse request_headers = 1;

    // The server must send back this message in response to a message with the
    // "response_headers" field set.
    HeadersResponse response_headers = 2;

    // The server must send back this message in response to a message with
    // the "request_body" field set.
    BodyResponse request_body = 3;

    // The server must send back this message in response to a message with
    // the "response_body" field set.
    BodyResponse response_body = 4;

    // The server must send back this message in response to a message with
    // the "request_trailers" field set.
    TrailersResponse request_trailers = 5;

    // The server must send back this message in response to a message with
    // the "response_trailers" field set.
    TrailersResponse response_trailers = 6;

    // If specified, return the specified response downstream to the client
    // and end further filter processing on the request path.
    ImmediateResponse immediate_response = 7;
  }

  // Optional metadata that will be emitted as dynamic metadata to be consumed by the next
  // filter. This metadata will be placed in the namespace "envoy.filters.http.ext_proc".
  google.protobuf.Struct dynamic_metadata = 8;

  // Override how parts of the HTTP request and response are processed
  // for the duration of this particular request/response only. Servers
  // may use this to intelligently control how requests are processed
  // based on the headers and other metadata that they see.
  envoy.extensions.filters.http.ext_proc.v3alpha.ProcessingMode mode_override = 9;
}

// The following are messages that are sent to the server.

// This message is sent to the external server when the HTTP request and responses
// are first received.
message HttpHeaders {
  // The HTTP request headers. All header keys will be
  // lower-cased, because HTTP header keys are case-insensitive.
  config.core.v3.HeaderMap headers = 1;

  // The values of properties selected by the "request_properties"
  // list in the configuration. Each entry in the map is the
  // value of the specified property.
  map<string, google.protobuf.Struct> properties = 2;

  // If true, then there is no message body associated with this
  // request or response.
  bool end_of_stream = 3;
}

// This message contains the message body that Envoy sends to the external server.
message HttpBody {
  bytes body = 1;

  bool end_of_stream = 2;
}

// This message contains the trailers.
message HttpTrailers {
  config.core.v3.HeaderMap trailers = 1;
}

// The following are messages that may be sent back by the server.

// This message must be sent in response to a request_headers or response_headers
// message.
// [#next-free-field: 6]
message HeadersResponse {
  // Instructions on how to manipulate the headers
  HeaderMutation header_mutation = 1;

  // If set, add a body to the message if none was present
  // before.
  BodyMutation body_mutation = 2;

  // If set, add trailers to the message. If and only if this
  // field is combined with the end_stream flag, then
  // the trailers will be sent along with the request or
  // response -- otherwise, this field is ignored.
  config.core.v3.HeaderMap trailers = 3;

  // If true, then replace the request and response with the contents
  // of this message. If header_mutation is set, apply it to the
  // headers. If body_mutation is set and contains a body, then add that
  // body to the request or response, even if one does not already exist --
  // otherwise, clear the body. Any additional body and trailers
  // received from downstream will be ignored, and any additional messages
  // from the remote server on either the request or response path will
  // also be ignored for the duration of this request or response.
  bool end_of_stream = 4;

  // Clear the route cache for the current request.
  // This is necessary if the remote server
  // modified headers that are used to calculate the route.
  bool clear_route_cache = 5;
}

// This message must be sent in response to a request_trailers
// or response_trailers message.
message TrailersResponse {
  // Instructions on how to manipulate the trailers
  HeaderMutation header_mutation = 1;
}

// TODO this is now the same as HeaderResponse. However, do we think
// that will always be the case? Keeping it as a separate message
// for now.
// [#next-free-field: 6]
message BodyResponse {
  // If set, manipulate HTTP headers along with the body. Note that
  // this is only possible if the headers have not been sent down
  // the filter chain yet. Practically, then, this field will be ignored
  // unless the request or response BodySendMode is BUFFERED or
  // BUFFERED_PARTIAL, which means that Envoy is buffering the body
  // in memory while waiting for this response.
  HeaderMutation header_mutation = 1;

  // Request that the last body chunk delivered be modified.
  BodyMutation body_mutation = 2;

  // If set, add trailers to the message. If and only if this
  // field is combined with the end_stream flag, then
  // the trailers will be sent along with the request or
  // response -- otherwise, this field is ignored.
  config.core.v3.HeaderMap trailers = 3;

  // If true, then replace the request and response with the contents
  // of this message. If header_mutation is set and headers have not
  // already been forwarded down the filter chain, apply it to the
  // headers. If body_mutation is set and contains a body, then add that
  // body to the request or response, even if one does not already exist --
  // otherwise, clear the body. Any additional body and trailers
  // received from downstream will be ignored, and any additional messages
  // from the remote server on either the request or response path will
  // also be ignored for the duration of this request or response.
  bool end_of_stream = 4;

  // Clear the route cache for the current request.
  // This is necessary if the remote server
  // modified headers that are used to calculate the route.
  bool clear_route_cache = 5;
}

// This message completely replaces the response. If the response headers
// have not already been sent downstream, then this will cause the filter to
// immediately return the data specified here to the downstream
// system, stop further filter processing, and ignore further messages coming
// from the remote server. However, if the response headers have already been
// delivered downstream, then this message will be ignored.
message ImmediateResponse {
  type.v3.HttpStatus status = 1 [(validate.rules).message = {required: true}];

  config.core.v3.HeaderMap headers = 2;

  bytes body = 3;

  config.core.v3.HeaderMap trailers = 4;
}

// Change HTTP headers or trailers by appending, replacing, or removing
// headers.
message HeaderMutation {
  // Add or replace HTTP headers.
  repeated config.core.v3.HeaderValueOption set_headers = 1;

  // Remove these HTTP headers.
  repeated string remove_headers = 2;
}

// Replace the entire message body chunk received in the corresponding
// HttpBody message with this new body, or clear the body.
message BodyMutation {
  oneof mutation {
    // The entire body to replace
    bytes body = 1;

    // Clear the corresponding body chunk
    bool clear_body = 2;
  }
}
